File: src/main/java/cz/wa/secretagent/view/model/HealthBarModel.java
//package cz.wa.secretagent.view.model;
//
//import java.util.Arrays;
//import java.util.Collection;
//import java.util.HashSet;
//import java.util.Set;
//
//import org.apache.commons.lang.Validate;
//
//import cz.wa.secretagent.view.SamGraphics;
//import cz.wa.secretagent.view.TileId;
//import cz.wa.secretagent.view.texture.GLGraphics;
//import cz.wa.secretagent.view.texture.TextureToDraw;
//import cz.wa.wautils.math.Rectangle2D;
//
///**
// * Model of health bar.
// *
// * @author Ondrej Milenovsky
// */
//public class HealthBarModel extends AbstractModel {
//    private final TileId frameTileId;
//    private final TileId healthTileId;
//
//    private transient TextureToDraw frameTex;
//    private transient TextureToDraw healthTex;
//
//    public HealthBarModel(TileId frameTileId, TileId healthTileId, double scale) {
//        super(scale);
//        this.frameTileId = frameTileId;
//        this.healthTileId = healthTileId;
//    }
//
//    @Override
//    public ModelType getType() {
//        return ModelType.HEALTH_BAR;
//    }
//
//    @Override
//    public Set<TileId> getAllTileIds() {
//        return new HashSet<TileId>(Arrays.asList(frameTileId, healthTileId));
//    }
//
//    @Override
//    public boolean hasLinkedTextures() {
//        return healthTex != null;
//    }
//
//    @Override
//    protected Rectangle2D linkTexturesInternal(SamGraphics graphics) {
//        frameTex = ((GLGraphics) graphics).getTile(frameTileId);
//        healthTex = ((GLGraphics) graphics).getTile(healthTileId);
//        Validate.notNull(frameTex, "missing tile " + frameTileId);
//        Validate.notNull(healthTex, "missing tile " + healthTileId);
//        if (frameTex != null) {
//            return getModelBounds(frameTex.getTileBounds());
//        } else {
//            return null;
//        }
//    }
//
//    @Override
//    public Collection<TextureToDraw> getAllTextures() {
//        return Arrays.asList(frameTex, healthTex);
//    }
//
//    public TextureToDraw getFrameTexture() {
//        return frameTex;
//    }
//
//    public TextureToDraw getHealthTexture() {
//        return healthTex;
//    }
//
//    @Override
//    public int hashCode() {
//        final int prime = 31;
//        int result = super.hashCode();
//        result = prime * result + ((frameTileId == null) ? 0 : frameTileId.hashCode());
//        result = prime * result + ((healthTileId == null) ? 0 : healthTileId.hashCode());
//        return result;
//    }
//
//    @Override
//    public boolean equals(Object obj) {
//        if (this == obj) {
//            return true;
//        }
//        if (!super.equals(obj)) {
//            return false;
//        }
//        if (getClass() != obj.getClass()) {
//            return false;
//        }
//        HealthBarModel other = (HealthBarModel) obj;
//        if (frameTileId == null) {
//            if (other.frameTileId != null) {
//                return false;
//            }
//        } else if (!frameTileId.equals(other.frameTileId)) {
//            return false;
//        }
//        if (healthTileId == null) {
//            if (other.healthTileId != null) {
//                return false;
//            }
//        } else if (!healthTileId.equals(other.healthTileId)) {
//            return false;
//        }
//        return true;
//    }
//
//}

File: src/main/kotlin/secretAgent/view/model/HealthBarModel.kt
package secretAgent.view.model

import cz.wa.secretagent.view.SAMGraphics
import cz.wa.secretagent.view.TileId
import cz.wa.secretagent.view.model.AbstractModel
import cz.wa.wautils.math.Rectangle2D
import secretAgent.view.renderer.GLGraphics
import secretAgent.view.renderer.TextureToDraw
import secretAgent.world.ModelType
import java.util.*
import kotlin.collections.HashSet

/**
 * Model of health bar.
 *
 * @author Ondrej Milenovsky
 */
class HealthBarModel(
        private val frameTileId: TileId?,
        private val healthTileId: TileId?, scale: Double) : AbstractModel(scale) {

    @Transient
    var frameTexture: TextureToDraw? = null
        private set
    @Transient
    var healthTexture: TextureToDraw? = null
        private set

    override val type: ModelType
        get() = ModelType.HEALTH_BAR

    override val allTileIds: Set<TileId>
        get() {
            val res = HashSet<TileId>()
            frameTileId?.let { res += it }
            healthTileId?.let { res += it }
            return res
        }

    override fun hasLinkedTextures() = healthTexture != null

    override fun linkTexturesInternal(graphics: SAMGraphics): Rectangle2D? {
        frameTexture = (graphics as GLGraphics).getTile(frameTileId!!) ?: throw Error("missing tile $frameTileId")
        healthTexture = graphics.getTile(healthTileId!!) ?: throw Error("missing tile $healthTileId")
        return frameTexture?.let { AbstractModel.getModelBounds(it.tileBounds) }
    }

    // TODO nullability
    public override fun getAllTextures(): Collection<TextureToDraw> = Arrays.asList<TextureToDraw>(frameTexture, healthTexture)

    override fun hashCode(): Int {
        val prime = 31
        var result = super.hashCode()
        result = prime * result + (frameTileId?.hashCode() ?: 0)
        result = prime * result + (healthTileId?.hashCode() ?: 0)
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (!super.equals(other)) {
            return false
        }
        return other is HealthBarModel && frameTileId == other.frameTileId && healthTileId == other.healthTileId
    }
}